engine/execution/computation/computer/uploader/retryable_uploader.go

```diff
@@ -0,0 +1,245 @@
+	execDataService    state_synchronization.ExecutionDataService
+	unit               *engine.Unit
+	metrics            module.ExecutionMetrics
+	blocks             storage.Blocks
+	commits            storage.Commits
+	transactionResults storage.TransactionResults
+	uploadStatusStore  storage.ComputationResultUploadStatus
+}
+
+func NewBadgerRetryableUploader(
+	uploader *AsyncUploader,
+	blocks storage.Blocks,
+	commits storage.Commits,
+	transactionResults storage.TransactionResults,
+	uploadStatusStore storage.ComputationResultUploadStatus,
+	execDataService state_synchronization.ExecutionDataService,
+	metrics module.ExecutionMetrics) *BadgerRetryableUploader {
+
+	// check params
+	if uploader == nil {
+		log.Error().Msg("nil uploader passed in")
+		return nil
+	}
+	if blocks == nil || commits == nil || transactionResults == nil ||
+		uploadStatusStore == nil || execDataService == nil {
+		log.Error().Msg("not all storage parameters are valid")
+		return nil
+	}
+
+	// When Upload() is successful, the stored ComputationResult in BadgerDB will be removed.
+	onCompleteCB := func(computationResult *execution.ComputationResult, err error) {
+		if computationResult == nil {
+			log.Warn().Msg("nil ComputationResult parameter")
+			return
+		}
+
+		if err != nil {
+			log.Warn().Msgf("ComputationResults upload failed with ID %s",
+				computationResult.ExecutionDataID)
+			return
+		}
+
+		// Update upload status as Done(true)
+		if err := uploadStatusStore.Upsert(computationResult.ExecutionDataID,
+			true /*upload complete*/); err != nil {
+			log.Warn().Msgf(
+				"ComputationResults with ID %s failed to be updated on local disk. ERR: %s ",
+				computationResult.ExecutionDataID.String(), err.Error())
+		}
+
+		metrics.ExecutionComputationResultUploaded()
+	}
+
+	uploader.SetOnCompleteCallback(onCompleteCB)
+
+	return &BadgerRetryableUploader{
+		uploader:           uploader,
+		execDataService:    execDataService,
+		unit:               engine.NewUnit(),
+		metrics:            metrics,
+		blocks:             blocks,
+		commits:            commits,
+		transactionResults: transactionResults,
+		uploadStatusStore:  uploadStatusStore,
+	}
+}
+
+func (b *BadgerRetryableUploader) Ready() <-chan struct{} {
+	return b.uploader.Ready()
+}
+
+func (b *BadgerRetryableUploader) Done() <-chan struct{} {
+	return b.uploader.Done()
+}
+
+func (b *BadgerRetryableUploader) Upload(computationResult *execution.ComputationResult) error {
+	if computationResult == nil || computationResult.ExecutableBlock == nil {
+		return errors.New("ComputationResult or its ExecutableBlock is nil when Upload() is called.")
+	}
+
+	// Before upload we store ComputationResult upload status to BadgerDB as false before upload is done.
+	// It will be marked as true when upload completes.
+	if err := b.uploadStatusStore.Upsert(computationResult.ExecutionDataID, false /*not completed*/); err != nil {
+		log.Warn().Msgf("failed to store ComputationResult into local DB with ID %s",
```

Comment:
Shouldn't we bubble up this error instead of just logging?

Reply:
the reason for this non-blocking logging is that even when there's sth going on with BadgerDB ops, upload can still be executed